src/clients/MultiCallerClient.ts

```diff
@@ -41,28 +44,151 @@ export const unknownRevertReasonMethodsToIgnore = new Set([
+      // To be resolved once the legacy implementation is removed and the callers have been updated.
+      const txnHashes: { [chainId: number]: string[] } = await this.executeTxnQueues(simulate);
+      return Object.values(txnHashes).flat();
+    }
+
+    return this.executeTxnQueueLegacy(simulate);
+  }
+
+  // For each chain, collate the enqueued transactions and process them in parallel.
+  async executeTxnQueues(simulate = false): Promise<{ [chainId: number]: string[] }> {
+    const chainIds = [...new Set(Object.keys(this.valueTxns).concat(Object.keys(this.txns)))];
+
+    // One promise per chain for parallel execution.
+    const results = await Promise.allSettled(
+      chainIds
+        .filter((_chainId) => {
+          const chainId = Number(_chainId);
+          return this.txns[chainId]?.length > 0 || this.valueTxns[chainId]?.length > 0;
+        })
+        .map((_chainId) => {
+          const chainId = Number(_chainId);
+          const txns: AugmentedTransaction[] = this.txns[chainId];
+          const valueTxns: AugmentedTransaction[] = this.valueTxns[chainId];
+
+          this.clearTransactionQueue(chainId);
+          return this.executeChainTxnQueue(chainId, txns, valueTxns, simulate);
+        })
+    );
+
+    // Collate the results for each chain.
+    const txnHashes: { [chainId: number]: string[] } = Object.fromEntries(
+      results.map((result, idx) => {
+        const chainId = chainIds[idx];
+        if (isPromiseFulfilled(result)) return [chainId, result.value.map((txnResponse) => txnResponse.hash)];
+        assert(isPromiseRejected(result), `Unexpected multicall result status: ${result?.status ?? "unknown"}`);
+        return [chainId, [result.reason]];
+      })
+    );
+
+    return txnHashes;
+  }
+
+  // For a single chain, simulate all potential multicall txns and group the ones that pass into multicall bundles.
+  // Then, submit a concatenated list of value txns + multicall bundles. If simulation was requested, log the results
+  // and return early.
+  async executeChainTxnQueue(
+    chainId: number,
+    txns: AugmentedTransaction[] = [],
+    valueTxns: AugmentedTransaction[] = [],
+    simulate = false
+  ): Promise<TransactionResponse[]> {
+    const networkName = getNetworkName(chainId);
+
+    const nTxns = txns.length + valueTxns.length;
+    this.logger.debug({
+      at: "MultiCallerClient#executeTxnQueue",
+      message: `${simulate ? "Simulating" : "Executing"} ${nTxns} transaction(s) on ${networkName}.`,
+    });
+
+    const txnSims = await Promise.allSettled([
+      this.simulateTransactionQueue(txns),
+      this.simulateTransactionQueue(valueTxns),
+    ]);
+
+    let _txns: AugmentedTransaction[];
+    let _valueTxns: AugmentedTransaction[];
+    [_txns, _valueTxns] = txnSims.map((result) => {
+      return isPromiseFulfilled(result) ? result.value : ([] as AugmentedTransaction[]);
+    });
```

Comment:
nit: would this work?
```suggestion
    const [_txns, _valueTxns] = txnSims.map((result): AugmentedTransaction[]  => {
      return isPromiseFulfilled(result) ? result.value : [];
    });
```

Reply:
Nice suggestion - much more concise! Implemented here: https://github.com/across-protocol/relayer-v2/pull/424/commits/d2c7ed520e057c5f8bc23523af1bbb7bc7c98ca4.